core/dataflow/src/main/resources/udf_server_main.py

```diff
@@ -0,0 +1,185 @@
+
+	def do_put(self, context, descriptor, reader, writer):
+		"""
+		Pass Arrow stream from the client to the server. The data must be associated with a `FlightDescriptor`,
+		which can be either a path or a command. Here the path is not actually a path on the disk,
+		but rather an identifier.
+		"""
+		key = UDFServer.descriptor_to_key(descriptor)
+		self.flights[key] = reader.read_all()
+
+	def do_get(self, context, ticket):
+		"""
+		Before getting the stream, the client must first ask the server for available tickets
+		(to the specified dataset) of the specified `FlightDescriptor`.
+		"""
+		key = ast.literal_eval(ticket.ticket.decode())
+		if key not in self.flights:
+			print("Flight Server:\tNOT IN")
+			return None
+		return pyarrow.flight.RecordBatchStream(self.flights[key])
+
+	def do_action(self, context, action):
+		"""
+		Each (implementation-specific) action is a string (defined in the script). The client is expected to know
+		available actions. When a specific action is called, the server executes the corresponding action and
+		maybe will return any results, i.e. a generalized function call.
+		"""
+		if action.type == "healthcheck":
+			# to check the status of the server to see if it is running.
+			yield pyarrow.flight.Result(pyarrow.py_buffer(b'Flight Server is up and running!'))
+		elif action.type == "shutdown":
+			# to shutdown the server.
+			yield pyarrow.flight.Result(pyarrow.py_buffer(b'Flight Server is shut down!'))
+			# Shut down on background thread to avoid blocking current
+			# request
+			threading.Thread(target=self._shutdown).start()
+		elif action.type == "open":
+			# open UDF
+			user_args_table = self.flights[self.descriptor_to_key(pyarrow.flight.FlightDescriptor.for_path(b'args'))]
+			user_args_list = user_args_table.to_pydict()['args']
+			self.udf_op.open(user_args_list)
+			yield pyarrow.flight.Result(pyarrow.py_buffer(b'Success!'))
+		elif action.type == "compute":
+			# execute UDF
+			# prepare input data
+			input_key = self.descriptor_to_key(pyarrow.flight.FlightDescriptor.for_path(b'toPython'))
+			input_table = self.flights[input_key]  # type: pyarrow.Table
+			input_dataframe = input_table.to_pandas()  # type: pandas.DataFrame
+			output_data_list = []
+			# execute and get output data
+			for index, row in input_dataframe.iterrows():
+				# FIXME: not sure what nth_child should behave here
+				try:
+					self.udf_op.accept(row)
+				except:
+					print_exc()
+				while self.udf_op.has_next():
+					output_data_list.append(self.udf_op.next())
+			output_dataframe = pandas.DataFrame.from_records(output_data_list)
+			# send output data to Java
+			output_key = self.descriptor_to_key(pyarrow.flight.FlightDescriptor.for_path(b'FromPython'))
+			try:
+				self.flights[output_key] = pyarrow.Table.from_pandas(output_dataframe)
+			except ArrowTypeError:
+				print_exc()
+			self.flights.pop(input_key)
+			yield pyarrow.flight.Result(pyarrow.py_buffer(b'Success!'))
```

Comment:
Now we need to cleary define the interfaces between each request and response pair. 
For compute request, I suggested that we report the error message to the java side, and I'm not seeing it being implemented here. If an exception occurs during processing, it just simply print the error message and moves on. This is not the expected behavior, if the python operator throws an exception, the entire workflow should fail.

I suggest we define the response of "compute" request to be a JSON object like:
```
status: (success or fail)
errorMessage: (only exists on failure)
```

Reply:
I'll make exception handling more complete.

Reply:
I've changed the response format of "compute". Now when an exception happens in compute, it will be displayed.
In the following demo, I deliberately raise an exception in demo_map_udf.py.
![image](https://user-images.githubusercontent.com/36582710/89517522-6097f500-d80c-11ea-922a-4c00823795a3.png)